src/ImageSharp/Formats/ImageDecoder.cs

```diff
@@ -0,0 +1,246 @@
+
+            // Issue #2259. Our buffered reads may have left the stream in an incorrect non-zero position.
+            // Reset the position of the seekable stream if we did not read to the end to allow additional reads.
+            if (stream.CanSeek && stream.Position != s.Position && s.Position != s.Length)
+            {
+                stream.Position = position + s.Position;
+            }
+
+            return result;
+        }
+
+        if (stream.CanSeek)
+        {
+            return Action(stream, stream.Position);
+        }
+
+        Configuration configuration = options.Configuration;
+        using ChunkedMemoryStream memoryStream = new(configuration.MemoryAllocator);
+        stream.CopyTo(memoryStream, configuration.StreamProcessingBufferSize);
+        memoryStream.Position = 0;
+
+        return Action(memoryStream, 0);
+    }
+
+    internal static async Task<T> WithSeekableStreamAsync<T>(
+        DecoderOptions options,
+        Stream stream,
+        Func<Stream, CancellationToken, T> action,
+        CancellationToken cancellationToken)
+    {
+        Guard.NotNull(options, nameof(options));
+        Guard.NotNull(stream, nameof(stream));
+
+        if (!stream.CanRead)
+        {
+            throw new NotSupportedException("Cannot read from the stream.");
+        }
+
+        T Action(Stream s, long position, CancellationToken ct)
+        {
+            T result = action(s, ct);
+
+            // Issue #2259. Our buffered reads may have left the stream in an incorrect non-zero position.
+            // Reset the position of the seekable stream if we did not read to the end to allow additional reads.
+            if (stream.CanSeek && stream.Position != s.Position && s.Position != s.Length)
+            {
+                stream.Position = position + s.Position;
+            }
+
+            if (ct.IsCancellationRequested)
+            {
+                throw new TaskCanceledException();
+            }
+
+            return result;
+        }
+
+        // NOTE: We are explicitly not executing the action against the stream here as we do in WithSeekableStream() because that
+        // would incur synchronous IO reads which must be avoided in this asynchronous method. Instead, we will *always* run the
+        // code below to copy the stream to an in-memory buffer before invoking the action.
+        if (stream is MemoryStream ms)
+        {
+            return Action(ms, ms.Position, cancellationToken);
+        }
+
+        if (stream is ChunkedMemoryStream cms)
+        {
+            return Action(cms, cms.Position, cancellationToken);
+        }
+
+        if (stream is BufferedReadStream brs && brs.BaseStream is MemoryStream)
+        {
+            return Action(brs, brs.Position, cancellationToken);
+        }
+
+        if (stream is BufferedReadStream brs2 && brs2.BaseStream is ChunkedMemoryStream)
+        {
+            return Action(brs2, brs2.Position, cancellationToken);
+        }
```

Comment:
You're right! `WebPDecoder` was being called by `WebpTiffCompression` which would have caused this, but I've since refactored it to use `WebPDecoderCore` so can remove.